tests/posix/pthread_key/src/pthread_key.c

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <ztest.h>
+#include <kernel.h>
+#include <posix/pthread.h>
+
+#define STACKSZ 1024
+#define BUFFSZ 48
+
+K_THREAD_STACK_ARRAY_DEFINE(stacks, 1, STACKSZ);
+
+pthread_key_t key;
+static pthread_once_t key_once;
+
+void *thread_top(void *p1)
+{
+	int ret = -1;
+
+	void *value;
+	void *getval;
+	char *buffer[BUFFSZ];
+
+	zassert_true((int)(value = k_malloc(sizeof(buffer))),
+			"thread could not allocate storage\n");
+
+	ret = pthread_setspecific(key, (void *)value);
+
+	/* TESTPOINT: Check if thread's value is associated with key */
+	zassert_false(ret, "pthread_setspecific failed\n");
+
+	getval = 0;
+
+	getval = pthread_getspecific(key);
+
+	/* TESTPOINT: Check if pthread_getspecific returns the same value
+	 * set by pthread_setspecific
+	 */
+	zassert_equal(value, getval,
+			"set and retrieved values are different\n");
+
+	printk("\nset value = %d and retrieved value = %d\n\n",
+			(int)value, (int)getval);
+
+	ret = pthread_key_delete(key);
+
+	/* TESTPOINT: Check if key is deleted */
+	zassert_false(ret, "attempt to delete key failed");
+
+	return NULL;
+}
+
+
+static void make_key(void)
+{
+	(void) pthread_key_create(&key, NULL);
```

Comment:
Please also test the creation of multiple keys, and make sure that pthread_getspecific() and pthread_setspecific() will set/get the values for multiple keys.  Also perform negative testing (e.g. pthread_getspecific(key1) should not return any values set with pthread_setspecific(key2, ...)).

Reply:
Also, tests should never cast the result of something that can fail to `(void)`.  Ensure that you can make pthread_key_create() fail and test for that.

Reply:
Yes, I was planning to add two more functions to this test case once this patch gets merged - for multiple threads using same key, and same thread using multiple keys complete with negative tests. I will add this test later this week in M3, I will remove the cast here and check for failure for now. Hope this is okay.